Source/WebCore/rendering/RenderListMarker.cpp

```diff
@@ -2025,9 +2001,69 @@ StringView RenderListMarker::textWithoutSuffix() const
     return StringView { m_textWithSuffix }.left(m_textWithoutSuffixLength);
 }
 
-CSSCounterStyle* RenderListMarker::counterStyle() const
+RefPtr<CSSCounterStyle> RenderListMarker::counterStyle() const
+{
+    return document().counterStyleRegistry().resolvedCounterStyle(style().listStyleType().identifier);
+}
+
+bool RenderListMarker::markerNeedsWidthAdjust() const
+{
+    auto listType = style().listStyleType().type;
+    if (listType == ListStyleType::Type::Disc || listType == ListStyleType::Type::Circle || listType == ListStyleType::Type::Square)
+        return true;
+
+    if (listType != ListStyleType::Type::CounterStyle)
+        return false;
+    auto counter = counterStyle();
+    return counter && (counter->isCircle() || counter->isDisc() || counter->isSquare());
+}
+
+bool RenderListMarker::isDisc() const
+{
+    auto listType = style().listStyleType().type;
+    if (listType == ListStyleType::Type::Disc)
+        return true;
+
+    if (listType != ListStyleType::Type::CounterStyle)
+        return false;
+    auto counter = counterStyle();
+    return counter && counter->isDisc();
+}
+
+bool RenderListMarker::isCircle() const
+{
+    auto listType = style().listStyleType().type;
+    if (listType == ListStyleType::Type::Circle)
+        return true;
+
+    if (listType != ListStyleType::Type::CounterStyle)
+        return false;
+    auto counter = counterStyle();
+    return counter && counter->isCircle();
+}
+
+bool RenderListMarker::isSquare() const
+{
+    auto listType = style().listStyleType().type;
+    if (listType == ListStyleType::Type::Square)
+        return true;
+
+    if (listType != ListStyleType::Type::CounterStyle)
+        return false;
+    auto counter = counterStyle();
+    return counter && counter->isSquare();
+}
+
+bool RenderListMarker::isDisclosureClosed() const
 {
-    return document().counterStyleRegistry().resolvedCounterStyle(style().listStyleType().identifier).get();
+    auto listType = style().listStyleType().type;
+    if (listType == ListStyleType::Type::DisclosureClosed)
+        return true;
+
+    if (listType != ListStyleType::Type::CounterStyle)
+        return false;
+    auto counter = counterStyle();
+    return counter && counter->isDisclosureClosed();
 }
```

Comment:
These methods should probably be on ListStyleType, rather than CounterStyle and RenderListMarker

Reply:
That is very true, thanks!